### PR TITLE
Minor improvements install instructions and documentation

### LINKED ISSUE #407: error: Cannot compile MPI programs. Check your configuration!!!
Hi there,

when installing mpi4py from PyPI (`pip install mpi4py`) users get the error `error: Cannot compile MPI programs. Check your configuration!!!` if they do not have a working MPI implementaion installed. There is no any warning or note both on README and on mpi4py's PyPI page as to the required prerequisites. I was wondering if we could add a note/warning regarging installation of mpi4py from PyPI, which says something like `you have to have a working MPI implementation before installing mpi4py`.

### PR DESCRIPTION
**Pull request by dalcinl**
Fixes #407.

cc: @YarShev @leofang

### CODE REVIEW

File: conf/mpidistutils.py [hunk 1]
```diff
@@ -908,13 +908,38 @@ def configure_mpi(ext, config_cmd):
...
+    if sys.platform == 'linux':
+        errmsg += [
+            "You may be missing the MPICH or Open MPI devel package:",
+            "* On Fedora/RHEL:",
+            "  $ sudo dnf install mpich-devel     # for MPICH",
+            "  $ sudo dnf install openmpi-devel   # for Open MPI",
+            "* On Debian/Ubuntu:",
+            "  $ sudo apt install libmpich-dev    # for MPICH",
+            "  $ sudo apt install libopenmpi-dev  # for Open MPI",
+        ]
```
Comment (leofang) [inline conf/mpidistutils.py:924]:
```suggestion
            "  $ sudo apt install libopenmpi-dev  # for Open MPI",
            "* Or check with your system administrator / refer to the user guide if using a cluster."
        ]
```

File: conf/mpidistutils.py [hunk 2]
```diff
@@ -908,13 +908,38 @@ def configure_mpi(ext, config_cmd):
...
+    if sys.platform == 'darwin':
+        errmsg += [
+            "Install MPICH or Open MPI with Homebrew or MacPorts:"
+            "  $ brew install mpich|openmpi  # Homebrew",
+            "  $ port install mpich|openmpi  # MacPorts",
+        ]
```
Comment (leofang) [inline conf/mpidistutils.py:930]: Q: Why is this needed for macOS? Didn't both of them provide binary packages?
- https://formulae.brew.sh/formula/mpi4py
- https://ports.macports.org/port/py-mpi4py/details/

So same as with conda, we shouldn't need special instruction?

File: README.rst [hunk 3]
```diff
@@ -25,8 +25,8 @@ Interface* (`MPI <https://www.mpi-forum.org/>`_) standard. It is
 implemented on top of the MPI specification and exposes an API which
 grounds on the standard MPI-2 C++ bindings.
 
-Dependencies
-------------
+Prerequisites
+-------------
```
Comment (leofang) [inline README.rst:29]: I have to say this is really like babysitting, I see no value added, but now that you've made the change I wouldn't even spend a second to argue 🙂

**Review by leofang — APPROVED** [review]

File: conf/mpidistutils.py (hunk 1 above)
Comment (dalcinl) [reply conf/mpidistutils.py:924]: And Leo Fang just proved my point... whatever you write, is never enough, there is always something extra to say...

File: conf/mpidistutils.py (hunk 2 above)
Comment (leofang) [reply conf/mpidistutils.py:930]: Fair enough, let me add more code suggestions.

File: README.rst (hunk 3 above)
Comment (dalcinl) [reply README.rst:29]: The new word sounds more pretentious... isn't that enough of a good reason for the change?

File: conf/mpidistutils.py (hunk 2 above)
Comment (leofang) [reply conf/mpidistutils.py:930]:
```suggestion
    if sys.platform == 'darwin':
        errmsg += [
            "Install MPICH or Open MPI with Homebrew, MacPorts, or Conda:"
            "  $ brew install mpich|openmpi                  # Homebrew",
            "  $ port install mpich|openmpi                  # MacPorts",
            "  $ conda install -c conda-forge mpich|openmpi  # Conda",
        ]
```

File: conf/mpidistutils.py (hunk 1 above)
Comment (leofang) [reply conf/mpidistutils.py:924]: (Suggestion updated.)

Comment (dalcinl) [reply conf/mpidistutils.py:927]: @leofang But your suggestion is wrong, you are missing the `{mpich|openmpi}-mpicc` package to pull in the compiler package, otherwise things will not work. Let me try to address it.

Comment (leofang) [reply conf/mpidistutils.py:927]: IIRC didn't we have bug reports for the mpicc wrappers from conda-forge?

File: conf/mpidistutils.py [hunk 4]
```diff
@@ -908,13 +908,44 @@ def configure_mpi(ext, config_cmd):
...
+    if sys.platform == 'win32':
+        errmsg += [
+            "Please install *Intel MPI* or *Microsoft MPI*."
+        ]
```
Comment (leofang) [inline conf/mpidistutils.py:940]: I think msmpi is also available from conda-forge, not sure about intel mpi, though

File: conf/mpidistutils.py (hunk 1 above)
Comment (dalcinl) [reply conf/mpidistutils.py:930]: Maybe, but I don't recall the context exactly. The fact is, the `mpicc` compiler wrapper and friends have the compiler command hardwired to the one used to build the MPI package. The <mpi>-mpicc packages are just convenience matapackages to install MPI + compiler toolchain. Without that, users get a confusing "command not found" error message.

File: conf/mpidistutils.py (hunk 2 above)
Comment (dalcinl) [reply conf/mpidistutils.py:936]: If we keep this way we could very well dump the full install instructions. This is why I don't link writing documentation: no matter how much effort you put into it, someone will always have something to say about it, it is never enough.

File: conf/mpidistutils.py (hunk 4 above)
Comment (dalcinl) [reply conf/mpidistutils.py:940]: Yes, not Intel MPI is also available. But think we are making a mistake mentioning conda-forge for the `pip install case`. An experienced user know what he/she is doing, a novice/beginner should probably stay away from mixing pip packages with conda packages.

File: conf/mpidistutils.py (hunk 2 above)
Comment (leofang) [reply conf/mpidistutils.py:936]: Exactly. I am hoping to use this PR to justify to @yarshev that this effort really makes no sense. Read the documentation, we can't be held responsible for MPI vendors.

File: conf/mpidistutils.py (hunk 1 above)
Comment (leofang) [reply conf/mpidistutils.py:930]: IIRC the problem was that conda-forge compiler does not play well, and if we bring in compiler toolchain from conda, instead of using the system one and just manually setting up `-I` and `-L`, we could run into issues.

**Comment by dalcinl** [comment]
@leofang  On  a second though, I believe is better to remove references to conda and conda-forge. Let's assume that `pip` users are not installing mpi4py in a conda environment, and if they do, they know about `conda install mpi4py`, and they choose to use pip for a good reason, and therefore they know what they are doing. Isn't that reasonable?

File: conf/mpidistutils.py (hunk 4 above)
Comment (leofang) [reply conf/mpidistutils.py:940]: Wouldn't this be the same case for brew/port?

**Comment by leofang** [comment]
I think the whole PR should just be nuked and replaced by a big banner:
```
mpi4py does not distribute pre-compiled binaries (wheels) on PyPI, when you do
"""
pip install mpi4py
"""
it would download the mpi4py sdist (source distribution) and build from source.
For pre-compiled binaries, please check out the documentation of your favorite package manager (HomeBrew/MacPorts/Conda/apt/yum/...). 
```
I think it's what confused @yarshev.

File: conf/mpidistutils.py (hunk 4 above)
Comment (dalcinl) [reply conf/mpidistutils.py:940]: Well, yes, you have a point. Now that I've spent so much time in it, let me do the full think at least for Linux and macOS. As usual, Windows may require more careful considerations I'm not willing to make for free.

File: conf/mpidistutils.py (hunk 1 above)
Comment (dalcinl) [reply conf/mpidistutils.py:930]: So let's remove all references to conda then. I don't want a clarification that contributes to spread even more confusion.

**Review by YarShev — COMMENTED** [review]
I think this looks much clearer now. I tried to install mpi4py from your branch.

```bash
pip install -e .
```

Got this error.

```bash
      error: Cannot compile MPI programs. Check your configuration!!!
      Installing mpi4py requires a working MPI implementation.
      If you are running on a supercomputer or cluster, check with
      the system administrator or refer to the system user guide.
      Otherwise, if you are running on a laptop or desktop computer,
      your may be missing the MPICH or Open MPI development package:
      * On conda environments with conda-forge packages, run:
        $ conda install -c conda-forge mpich-mpicc   # for MPICH
        $ conda install -c conda-forge openmpi-mpicc # for Open MPI
      * On Fedora/RHEL systems, run:
        $ sudo dnf install mpich-devel     # for MPICH
        $ sudo dnf install openmpi-devel   # for Open MPI
      * On Debian/Ubuntu systems, run:
        $ sudo apt install libmpich-dev    # for MPICH
        $ sudo apt install libopenmpi-dev  # for Open MPI
```

Based on the output I installed `mpich-mpicc`.

```bash
mamba install -c conda-forge mpich-mpicc
```

After that I was able to install mpi4py and ran a simple test.

The changes on `INSTALL.rst` also look good to me.

Many thanks! I hope this will help users a lot in installing mpi4py and depending on it packages.

**Comment by dalcinl** [comment]
> Based on the output I installed `mpich-mpicc`.

Well, I eventually removed that stuff. If you are using conda, you should just `conda install mpi4py`. 
Users wanting to mix conda packages and pip packages are expected to know what they are doing.

**Comment by YarShev** [comment]
Even after removing all references to conda, I think the changes are very valuable and much clearer for users.

**Review by YarShev — APPROVED** [review]

**Comment by YarShev** [comment]
@dalcinl, could you shed light on which release will contain the changes? Is there some cadence for mpi4py releases?

**Comment by dalcinl** [comment]
> could you shed light on which release will contain the changes? Is there some cadence for mpi4py releases?

See #386.

**Comment by YarShev** [comment]
@dalcinl, I was wondering if we should add some lines about installation on Windows if possible?

```
      * On Fedora/RHEL systems, run:
        $ sudo dnf install mpich-devel     # for MPICH
        $ sudo dnf install openmpi-devel   # for Open MPI
      * On Debian/Ubuntu systems, run:
        $ sudo apt install libmpich-dev    # for MPICH
        $ sudo apt install libopenmpi-dev  # for Open MPI
```

**Comment by dalcinl** [comment]
@YarShev I mean, why the `"Please install *Intel MPI* or *Microsoft MPI*." message is not enough?

**Comment by YarShev** [comment]
@dalcinl, I wonder if there is a way (one-line command) to install an MPI implementation on Windows using some kind of a package manager for Windows as for Linux? We could put that line in the error message above.

**Comment by dalcinl** [comment]
* There are many ways to install packages on Windows, from native installers you download from the MPI provider webpages directly, to other pkg managers like winget, nuget, chocolately, just to name the ones I've heard about. I'm not a Windows user, and I don not have enough expertise as to suggest others what they should use. Moreover, given that there are so many options, I'm not willing to dump a complete MPI installation manual just for the occasional clueless user. 
* Installing MPI is just one piece of the puzzle, users should install a MSVC, wich is a large download, and I'm not even sure it is available without subscribing to some MS developer account. Without a C compiler, building mpi4py from sources would fail miserably. Am I supposed to teach users how to deal with that mess, too? Where do we stop?
* For a some time, mpi4py has provided pre-built binary wheels built against Microsoft MPI. That may change for next release, as Intel MPI is also available in Windows, and provides support for the MPI 3.1 standard. Long story short, this means that `pip install mpi4py` on Windows will install a pre-built wheel, and there is no need to build from sources. Of course, installing Microsoft MPI (or Intel MPI in the future) is still a requirement.
* Finally, if you are still worried that your Windows consumers will struggle with MPI, maybe we should point them to the stuff [here](https://github.com/mpi4py/mpi4py-publish) and provide the one-line instruction about how to pip install Intel MPI and mpi4py (that is, the "(i) Note" box).

**Comment by YarShev** [comment]
@dalcinl, I understand. Thank you for the comprehensive explanation. It seems to me that is a good point to add a link to the stuffe you mentioned in the error message - `"Please install Intel MPI or Microsoft MPI.". Of course, if it will not bloat up the error message.

**Comment by leofang** [comment]
> Am I supposed to teach users how to deal with that mess, too? Where do we stop?

[My position on this](https://github.com/mpi4py/mpi4py/pull/410#issuecomment-1760242439) has been that
> I think the whole PR should just be nuked and replaced by a big banner

**Comment by dalcinl** [comment]
This is a project of worldwide reach for a niche need with a very small dev team and zero funding. Every second I spend on new developments and maintenance is time I could use for tangible benefit in other activities. Yet, I do this work for personal fulfillment, to keep myself up-to-date and learn new thing, and to interact with professionals that know what they are doing and producing stuff many others benefit, either directly or indirectly. In this scheme of things, you like it or not, newbies have very low priority.

**Comment by YarShev** [comment]
Makes sense, we can leave the error message as is.
